### PR TITLE
efi/preinstall Add initial RunChecksContext API

### PR DESCRIPTION
**Pull request by chrisccoulson**
This API is easier to use than the other exported API (RunChecks) and
converts errors into error kinds that are suitable for representing in
JSON.

This API will eventually map certain error kinds to actions that can be
performed by this package in order to resolve some issues. The intention
here is that the installer UI eventually gives the user some information
about why FDE cannot be enabled, and it gives them the option of choosing
to fix some issues where it is possible and appropriate to do so (eg, say
the TPM device is detected but disabled, there should be an action to
enable it via the PPI).

The Action type and some pseudo actions are already defined - these are
actions that cannot be performed by this package, but are hints for
actions that can be performed by the caller, eg, shutdown, reboot and
reboot to firmware settings are some examples.

RunChecksContext will keep track of which actions are expected (based on
the previous response) and which actions are enabled - some actions are
enabled by default but may become disabled later on because of an error,
and some other actions will only be made available after performing
tests that indicate the action can be performed.

### CODE REVIEW

File: efi/preinstall/checks_context.go [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+	env          internal_efi.HostEnvironment
+	flags        CheckFlags
+	loadedImages []secboot_efi.Image
+	profileOpts  PCRProfileOptionsFlags
+
+	errs    []error
+	lastErr error
+	result  *CheckResult
+
+	availableActions map[Action]bool
```
Comment (pedronis) [inline efi/preinstall/checks_context.go:142]: this probably merits a comment as it seems it can store both true and false entries which is not always the case with such maps, sometimes they are really meant to be sets (so either no entry or true)

File: efi/preinstall/checks_context.go [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+
+// isActionExpected determines if the supplied action is an expected
+// response to the last call to [RunChecksContext.Run].
+func (c *RunChecksContext) isActionExpected(action Action) bool {
+	for _, expected := range c.expectedActions {
+		if expected == action {
+			return true
+		}
+	}
+	return false
```
Comment (pedronis) [inline efi/preinstall/checks_context.go:222]: tests don't seem to hit the false case

File: efi/preinstall/checks_context.go [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+		// unless we discover that they aren't later on.
+		availableActions: map[Action]bool{
+			ActionNone:               true,
+			ActionReboot:             true,
+			ActionShutdown:           true,
+			ActionRebootToFWSettings: true,
+			ActionContactOEM:         true,
+			ActionContactOSVendor:    true,
+		},
+		expectedActions: []Action{ActionNone},
```
Comment (pedronis) [inline efi/preinstall/checks_context.go:175]: do we need to set this as is also set in Run?

**Review by pedronis — COMMENTED** [review]
did a first pass, didn't look at tests yet

File: efi/preinstall/checks_context.go [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+
+// Run will run the specified action, and if that completes successfully will run another
+// iteration of [RunChecks] and test the result against the preferred [WithAutoTCGPCRProfile]
+// configuration. On success, this will return the CheckResult. On failure, this will return
+// an error which will either be a single ErrorKindAndActions, or multiple ErrorKindAndActions
+// wrapped by an error type that implements the [CompoundError] interface. If there are any
+// actions associated with an error, the install environment may try one or more of them in
+// order to try to resolve the issue that caused the error. In some cases, it may be appropriate
+// to ask permission from the user to perform an action.
+func (c *RunChecksContext) Run(ctx context.Context, action Action, args ...any) (*CheckResult, error) {
```
Comment (ernestl) [inline efi/preinstall/checks_context.go:478]: Should we consider returning a compound error always instead of mixing it up, to not require inspection to determine what we are dealing with?

File: efi/preinstall/errors.go [hunk 5]
```diff
@@ -814,3 +823,25 @@ func (e *UnsupportedRequiredPCRsError) Error() string {
 		return fmt.Sprintf("PCRs %v are required, but are unsupported", e.PCRs)
 	}
 }
+
+// ErrorKindAndActions is an error type that can be serialized to JSON, represented by
+// an error kind, associated arguments and a set of potential remedial actions.
+type ErrorKindAndActions struct {
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/errors.go:829]: That sounds ok with me.

File: efi/preinstall/errors.go [hunk 6]
```diff
@@ -814,3 +823,25 @@ func (e *UnsupportedRequiredPCRsError) Error() string {
...
+	ErrorKind ErrorKind       `json:"kind"`    // The error kind
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/errors.go:830]: I think it's ok for me to just rename these.

File: efi/preinstall/actions.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	ActionContactOSVendor Action = "contact-os-vendor"
+)
+
+// IsPseudoAction will return true if the action cannot actually be executed by
+// [RunChecksContext.Run], but the action is expected to be performed by the caller
+// (eg, snapd or the installer) instead.
+//
+// TODO: Add extra actions that can be performed by this package by passing the
+// action to [RunChecksContext.Run].
+func (a Action) IsPseudoAction() bool {
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/actions.go:76]: Yes, that's ok - I wasn't sure of what name to use here.

File: efi/preinstall/checks.go [hunk 8]
```diff
@@ -270,7 +270,9 @@ func RunChecks(ctx context.Context, flags CheckFlags, loadedImages []secboot_efi
 
 	logResults, err := checkFirmwareLogAndChoosePCRBank(tpm, log, mandatoryPcrs, checkLogFlags)
 	switch {
-	case tpm2.IsTPMError(err, tpm2.AnyErrorCode, tpm2.AnyCommandCode):
+	case tpm2.IsTPMError(err, tpm2.AnyErrorCode, tpm2.AnyCommandCode) ||
+		tpm2.IsTPMWarning(err, tpm2.AnyWarningCode, tpm2.AnyCommandCode) ||
+		isInvalidTPMResponse(err) || isTPMCommunicationError(err):
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks.go:275]: There aren't - I can probably add some more tests for this.

There are a few known cases of missing test coverage, but I have left some notes throughout the test files and in checks_context.go where I know test coverage is missing - it's mostly just for the harder-to-test edge cases and I'm going to fill in the gaps (maybe as a follow-up PR)

File: efi/preinstall/checks_context.go [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+	if errors.Is(err, ErrPreOSVerificationUsingDigests) {
+		return ErrorKindPreOSDigestVerificationDetected, nil, nil
+	}
+
+	var sbPcrErr *SecureBootPolicyPCRError
+	if errors.As(err, &sbPcrErr) {
+		return ErrorKindPCRUnusable, []any{internal_efi.SecureBootPolicyPCR}, nil
+	}
+
+	return ErrorKindInternal, nil, nil
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:423]: Yes, I need to figure out the best way of generating an error that gets here. I'll take a look.

File: efi/preinstall/checks_context.go (hunk 3 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:175]: It's reset in `Run` after running the supplied action - it's set initially here to bootstrap it otherwise `Run` would fail with an unexpected action. It might be better to just clear `expectedActions` and have `ActionNone` as implicitly accepted instead.

File: efi/preinstall/error_kinds.go [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+ */
+
+package preinstall
+
+// ErrorKind describes an error detected during preinstall checks when
+// using the [RunChecksContext] API.
+type ErrorKind string
+
+const (
+	// ErrorKindNone indicates that no error occurred.
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/error_kinds.go:27]: Ah, I thought I'd done this - I'll have another scan through to see where I've missed them.

File: efi/preinstall/checks_context.go [hunk 11]
```diff
@@ -0,0 +1,585 @@
...
+		for _, prop := range []tpm2.Property{tpm2.PropertyLockoutCounter, tpm2.PropertyLockoutInterval} {
+			val, err := tpm.GetCapabilityTPMProperty(prop)
+			if err != nil {
+				return ErrorKindNone, nil, fmt.Errorf("cannot read property %d: %w", prop, err)
+			}
+			vals = append(vals, val)
+		}
+		lockoutCounter = vals[0]
+		lockoutInterval = vals[1]
+		return ErrorKindTPMDeviceLockout, []any{time.Duration(lockoutInterval) * time.Second, time.Duration(lockoutInterval) * time.Second * time.Duration(lockoutCounter)}, nil
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:287]: That's fine with me.

File: efi/preinstall/checks_context.go [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/canonical/go-tpm2"
+	secboot_efi "github.com/snapcore/secboot/efi"
+	internal_efi "github.com/snapcore/secboot/internal/efi"
+)
+
+var errorKindToActions map[ErrorKind][]Action
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:34]: I can add some comments, but it's an OR rather than a sequence. There aren't any cases where a sequence of actions is required. There might be more than one error - each with an OR of associated actions, in which case a single action from each error can be executed, but they can be executed in any order.

File: efi/preinstall/checks_context.go [hunk 13]
```diff
@@ -0,0 +1,585 @@
...
+// should match the flags later passed to this function when creating a PCR profile. The PCRs that are
+// determined to be required to build the profile using [WithAutoTCGPCRProfile] will be made mandatory
+// automatically by this function by passing the relevant flags to [RunChecks].
+//
+// As [PCRProfileOptionsFlags] is intended to provide some limited amount of user customisation, this
+// API [RunChecksContext] should be executed again with the new set of flags, should the user wish to
+// change them. In this case, the caller should pass the PostInstallChecks flag as an initial flag.
+//
+// There is no need for the caller to supply any of these *SupportRequired flags as the initial flags,
+// and this may have the effect of limiting the number of devices which pass the checks.
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:158]: That is ok to me.

File: efi/preinstall/checks_context.go (hunk 1 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:142]: This is intended to have 3 states - empty implies that we need to test that the action is available, false is "tested, but unavailable" and true is "available".

File: efi/preinstall/checks_context.go (hunk 2 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:222]: I can add a test for this.

File: efi/preinstall/checks_context.go [hunk 14]
```diff
@@ -0,0 +1,585 @@
...
+		var requiredPCRsErr *UnsupportedRequiredPCRsError
+		if !errors.As(profileErr, &requiredPCRsErr) {
+			return nil, &ErrorKindAndActions{
+				ErrorKind: ErrorKindInternal,
+				err:       fmt.Errorf("cannot test whether a PCR combination can be generated: %w", err),
+			}
+		}
+
+		// Make any PCRs we're lacking support for mandatory so that they end
+		// up being returned in the RunChecks error on the next iteration,
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:558]: Yes, I think that's the case.

File: efi/preinstall/checks_context.go [hunk 15]
```diff
@@ -0,0 +1,585 @@
...
+				c.flags |= SecureBootPolicyProfileSupportRequired
+			}
+		}
+	}
+
+	if c.result != nil {
+		return c.result, nil
+	}
+
+	return nil, joinErrors(errs...)
```
Comment (ernestl) [inline efi/preinstall/checks_context.go:584]: This is more a comment on joinErrors.Error(), and also taking into consideration the suggestion to always return a compound error:

From the snapd side, at the point where we receive the error we are firstly interested to know that a preinstall-check error was detected, without needing all the detail about possible multiple errors, with args and actions. 

Could we perhaps provide the compound error with a concise high level error message without all the details of multiple errors, args and actions?

More specifically, on the snapd side, we have the concept of a unavailable reason (short error message) and a separate field for slice of error information that is only exposed at API level. It would be convenient if the compound error [can be used for the unavailable reason as is](https://github.com/canonical/snapd/pull/15373/files#diff-8133fed769942c6fe6d9075fbd0253dffcba86bc3fc94b96e1a546c7f707a416R387).

**Comment by ernestl** [comment]
Hi @chrisccoulson, 

Regarding the topic of ensuring that the [existing preinstall check](https://github.com/canonical/snapd/blob/master/secboot/secboot_tpm.go#L88) is sufficiently covered by the new preinstall check.

(1) is efi secureboot enabled [read efi var](https://github.com/canonical/snapd/blob/master/secboot/secboot_tpm.go#L121)
(2) check if tpm available (connect)
(3) check if tpm enabled
(4) if mode == TPMProvisionFull, ensure [tpm is not locked out](https://github.com/canonical/snapd/blob/master/secboot/secboot_tpm.go#L109)

- 1 is not covered, but secboot has capability to read efi vars, so something to add.
- In previous discussion we said 2  and 3 likely sufficiently covered, to confirm.

However, we do not have equivalency in terms of considering `tpm mode`. 
I want to propose that we add `tpm mode`  as parameter for run-checks to allow for more granular checks.

File: efi/preinstall/checks.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks.go:275]: If it's ok, I'm going to add some test coverage here in a follow-up. Properly fixing this requires an update to go-tpm2 and I already have a PR with some changes in to accommodate this - it would be a bit easier to land these extra tests after that.

File: efi/preinstall/checks.go
Comment (ernestl) [inline efi/preinstall/checks.go:1]: Consider CheckFlags constant for default (0), same as what is available for PCRProfileOptionsFlags.

File: efi/preinstall/errors.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/errors.go]: I've renamed this now.

File: efi/preinstall/errors.go [hunk 16]
```diff
@@ -814,3 +823,25 @@ func (e *UnsupportedRequiredPCRsError) Error() string {
...
+
+	err error `json:"-"` // The original error. This is not serialized to JSON.
+}
+
+func (e *ErrorKindAndActions) Error() string {
+	data, err := json.Marshal(e)
+	if err != nil {
+		data = []byte(fmt.Sprintf("cannot serialize error: %v", err))
+	}
+	return fmt.Sprintf("%v %s", e.err, string(data))
```
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/errors.go]: I've removed the JSON from the String implementation.

File: efi/preinstall/checks_context.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:37]: I've added a doc comment to this now.

File: efi/preinstall/checks_context.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:165]: I've added a test case that tests this specific sequence now.

File: efi/preinstall/checks_context.go (hunk 1 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:149]: I've added a doc comment for this now.

File: efi/preinstall/checks_context.go (hunk 2 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:232]: I've added a test case with an invalid action name now.

File: efi/preinstall/checks_context.go (hunk 3 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go]: I've changed the semantics of `expectedActions` now so that `ActionNone` is always expected, even when set to nil.

File: efi/preinstall/errors.go
Comment (ernestl) [inline efi/preinstall/errors.go:1]: Multiple cases of ` wrapped wrapped`

File: efi/preinstall/error_kinds.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (chrisccoulson) [reply efi/preinstall/error_kinds.go:34]: I think each one of these with arguments is documented now and with a description of what the arguments look like.

**Comment by chrisccoulson** [comment]
> Hi @chrisccoulson,
> 
> Regarding the topic of ensuring that the [existing preinstall check](https://github.com/canonical/snapd/blob/master/secboot/secboot_tpm.go#L88) is sufficiently covered by the new preinstall check.
> 
> (1) is efi secureboot enabled [read efi var](https://github.com/canonical/snapd/blob/master/secboot/secboot_tpm.go#L121) (2) check if tpm available (connect) (3) check if tpm enabled (4) if mode == TPMProvisionFull, ensure [tpm is not locked out](https://github.com/canonical/snapd/blob/master/secboot/secboot_tpm.go#L109)
> 
> In a previous discussion we agreed:
> 
> * 1 is not covered, but secboot has capability to read efi vars, so something to add.
> * In previous discussion we said 2  and 3 likely sufficiently covered, to confirm.
> 
> However, we do not have equivalency in terms of considering `tpm mode`. I want to propose that we add `tpm mode` as parameter for run-checks to allow for more granular checks.

There are explicit tests already for 2-4. The only one there isn't an explicit test for is 1, although some tests do read EFI variables and go-efilib generates a fairly consistent error if efivarfs isn't available ([efi.ErrVarsUnavailable])

File: efi/preinstall/checks_context.go (hunk 4 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:480]: I need to have a bit more of a think about this one as it's a bit more complicated than I initially realized because [RunChecksContext] is keeping a log of returned errors and I was going to do this by adding a deferred function to do the wrapping.

File: efi/preinstall/errors.go
Comment (chrisccoulson) [reply efi/preinstall/errors.go:1]: These are fixed now.

File: efi/preinstall/checks_context.go (hunk 15 above)
Comment (chrisccoulson) [reply efi/preinstall/checks_context.go:579]: I'm not sure I've fully understood the suggestion here, but I'm not sure I'm keen on the idea of updating the `Error()` implementation to hide information about errors - a lot of information in these errors already isn't going to be serialized via `WithKindAndActionsError` and it would be useful for debugging in some cases to see the full error message in a log somewhere.

**Comment by ernestl** [comment]
Discussed with Chris the idea of also passing warnings to installer via snapd /v2/systems/{label} api.
The implication if we did it:
 - warnings would need to classified to kind
 - end up in same error type as errors with actions, but without args and actions

Chris advised that this is beyond what was intended for warnings, which is only to allow snapd to log it for debugging purposes, "as a tool for extracting test failures that don't otherwise prevent the install from continuing, so in that sense, they only really exist for debugging purposes".

@pedronis

File: efi/preinstall/checks_context.go (hunk 4 above)
Comment (ernestl) [reply efi/preinstall/checks_context.go:480]: This was a convenience request, it's ok as is.

File: efi/preinstall/checks.go
Comment (ernestl) [reply efi/preinstall/checks.go:1]: Thanks

File: efi/preinstall/errors.go
[the remark this answers is not in the document]
Comment (ernestl) [reply efi/preinstall/errors.go:1]: Thanks

Comment (ernestl) [reply efi/preinstall/errors.go:1]: Thanks
